kube/src/config/file_config.rs

```diff
@@ -140,15 +142,162 @@ impl Kubeconfig {
+        }
     }
+
+    /// Create `Kubeconfig` from `KUBECONFIG` environment variable.
+    /// Supports list of files to be merged. `KUBECONFIG` value must contain a valid unicode data.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `KUBECONFIG` value contains the NUL character.
+    pub fn from_env() -> Result<Option<Self>> {
+        match std::env::var(KUBECONFIG) {
+            Ok(value) => {
+                let paths = value
+                    .split(if cfg!(windows) { ';' } else { ':' })
+                    .filter(|s| !s.is_empty())
+                    .collect::<Vec<_>>();
+                if paths.is_empty() {
+                    return Ok(None);
+                }
+
+                let merged = paths.iter().try_fold(Kubeconfig::default(), |m, p| {
+                    Kubeconfig::read_from(p).and_then(|c| m.merge(c))
+                })?;
+                Ok(Some(merged))
+            }
+
+            Err(std::env::VarError::NotPresent) => Ok(None),
+            Err(std::env::VarError::NotUnicode(_)) => Err(ConfigError::InvalidKubeconfigEnv.into()),
+        }
+    }
+
+    /// Merge kubeconfig file according to the rules described in
+    /// <https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/#merging-kubeconfig-files>
+    ///
+    /// > Merge the files listed in the `KUBECONFIG` environment variable according to these rules:
+    /// >
+    /// > - Ignore empty filenames.
+    /// > - Produce errors for files with content that cannot be deserialized.
+    /// > - The first file to set a particular value or map key wins.
+    /// > - Never change the value or map key.
+    /// >   Example: Preserve the context of the first file to set `current-context`.
+    /// >   Example: If two files specify a `red-user`, use only values from the first file's `red-user`.
+    /// >            Even if the second file has non-conflicting entries under `red-user`, discard them.
+    fn merge(mut self, mut next: Kubeconfig) -> Result<Self> {
+        if self.kind.is_some() && next.kind.is_some() && self.kind != next.kind {
+            return Err(ConfigError::KindMismatch.into());
+        }
+        if self.api_version.is_some() && next.api_version.is_some() && self.api_version != next.api_version {
+            return Err(ConfigError::ApiVersionMismatch.into());
+        }
+
+        self.kind = self.kind.or(next.kind);
+        self.api_version = self.api_version.or(next.api_version);
+
+        match self.preferences.as_mut() {
+            Some(base_prefs) => {
+                if let Some(mut next_prefs) = next.preferences.take() {
+                    base_prefs.colors = base_prefs.colors.or(next_prefs.colors);
+
+                    match base_prefs.extensions.as_mut() {
+                        Some(mut base_exts) => {
+                            if let Some(next_exts) = next_prefs.extensions {
+                                append_new_named(&mut base_exts, next_exts, |x| &x.name);
```

Comment:
Maybe we shouldn't merge this? ("Never change the value or map key.")

Changes `preferences.extensions` value. Just update if `None`?

Reply:
Hm, really not sure about this. Don't know enough about whether or not adding extensions together makes sense.
It's probably not super important at the moment since it's more a kubectl option afaikt, and we are not doing anything with it. I'd just take the first preferences that's a Some and not worry about it for now.